File: asdar/objects/scope.zig
// this is an object because:
//    * function objects have a definition scope, and multiple functions may
//      have the same definition scope
//    * scopes can be created at runtime
//    * a scope has to be destroyed when no more functions defined in it exist
//
// from this you can see that scopes must be reference counted anyway, and
// objects already implement reference counting

const std = @import("std");
const builtins = @import("../builtins.zig");
const Interp = @import("../interp.zig").Interp;
const objtyp = @import("../objtyp.zig");
const Object = objtyp.Object;


pub const Data = struct {
    interp: *Interp,
    local_vars: []?*Object,
    parent_scopes: []*Object,    // micro-optimization, see getForLevel
    owns_local_vars: bool,

    fn initGlobal(interp: *Interp) !Data {
        const locals = try std.mem.dupe(interp.object_allocator, ?*Object, builtins.object_array[0..]);
        errdefer interp.object_allocator.free(locals);
        const scopes = try interp.object_allocator.alloc(*Object, 0);
        errdefer interp.object_allocator.free(scopes);

        for (locals) |obj| {
            obj.?.incref();
        }
        return Data{
            .interp = interp,
            .local_vars = locals,
            .parent_scopes = scopes,
            .owns_local_vars = true,
        };
    }

    fn initSub(parent: *Object, nlocals: u16) !Data {
        const parentdata = parent.data.ScopeData;

        const locals = try parentdata.interp.object_allocator.alloc(?*Object, nlocals);
        errdefer parentdata.interp.object_allocator.free(locals);
        for (locals) |*obj| {
            obj.* = null;
        }

        const parents = try parentdata.interp.object_allocator.alloc(*Object, parentdata.parent_scopes.len + 1);
        errdefer parentdata.interp.object_allocator.free(parents);
        std.mem.copy(*Object, parents, parentdata.parent_scopes);
        parents[parentdata.parent_scopes.len] = parent;
        for (parents) |obj| {
            obj.incref();
        }

        return Data{
            .interp = parentdata.interp,
            .local_vars = locals,
            .parent_scopes = parents,
            .owns_local_vars = true,
        };
    }

    pub fn destroy(self: Data, decref_refs: bool, free_nonrefs: bool) void {
        if (decref_refs) {
            if (self.owns_local_vars) {
                for (self.local_vars) |obj| {
                    if (obj != null) {
                        obj.?.decref();
                    }
                }
            }
            for (self.parent_scopes) |obj| {
                obj.decref();
            }
        }
        if (free_nonrefs) {
            if (self.owns_local_vars) {
                self.interp.object_allocator.free(self.local_vars);
            }
            self.interp.object_allocator.free(self.parent_scopes);
        }
    }
};

var type_value = objtyp.Type.init([]objtyp.Attribute { });
pub const typ = &type_value;

pub fn createGlobal(interp: *Interp) !*Object {
    const data = try Data.initGlobal(interp);
    errdefer data.destroy(true, true);
    return try Object.init(interp, typ, objtyp.ObjectData{ .ScopeData = data });
}

pub fn createSub(parent: *Object, nlocals: u16) !*Object {
    const data = try Data.initSub(parent, nlocals);
    errdefer data.destroy(true, true);
    return try Object.init(parent.data.ScopeData.interp, typ, objtyp.ObjectData{ .ScopeData = data });
}

pub fn getForLevel(scope: *Object, level: u16) *Object {
    const data = scope.data.ScopeData;
    var result: *Object = undefined;
    if (level == data.parent_scopes.len) {
        result = scope;
    } else {
        std.debug.assert(level < data.parent_scopes.len);
        result = data.parent_scopes[level];
    }

    result.incref();
    return result;
}

pub fn getLocalVars(scope: *Object) []?*Object {
    return scope.data.ScopeData.local_vars;
}

// to destroy the return value, decref items and free the array with the object allocator
pub fn getLocalVarsOwned(scope: *Object) []?*Object {
    std.debug.assert(scope.data.ScopeData.owns_local_vars);
    scope.data.ScopeData.owns_local_vars = false;
    return scope.data.ScopeData.local_vars;
}


test "scope object creating and getForLevel" {
    var interp: Interp = undefined;
    try interp.init();
    defer interp.deinit();

    const global_scope = try createGlobal(&interp);
    defer global_scope.decref();
    const file_scope = try createSub(global_scope, 3);
    defer file_scope.decref();

    std.testing.expect(global_scope.data.ScopeData.parent_scopes.len == 0);
    std.testing.expect(file_scope.data.ScopeData.parent_scopes.len == 1);
    std.testing.expect(file_scope.data.ScopeData.parent_scopes[0] == global_scope);

    const a = getForLevel(global_scope, 0);
    defer a.decref();
    const b = getForLevel(file_scope, 0);
    defer b.decref();
    const c = getForLevel(file_scope, 1);
    defer c.decref();
    std.testing.expect(a == global_scope);
    std.testing.expect(b == global_scope);
    std.testing.expect(c == file_scope);
}

File: asdar/src/objects/scope.c
#include "scope.h"
#include <assert.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include "../builtin.h"
#include "../interp.h"
#include "../objtyp.h"
#include "err.h"

static void destroy_scope(Object *obj, bool decrefrefs, bool freenonrefs)
{
	ScopeObject *scope = (ScopeObject *)obj;
	if (decrefrefs) {
		for (size_t i = 0; i < scope->nlocals; i++)
			if (scope->locals[i])
				OBJECT_DECREF(scope->locals[i]);

		if (scope->nparents != 0) OBJECT_DECREF(scope->parents[scope->nparents - 1]);
	}

	if (freenonrefs) {
		free(scope->locals);
		free(scope->parents);
	}
}

ScopeObject *scopeobj_newsub(Interp *interp, ScopeObject *parent, size_t nlocals)
{
	Object **locals;
	if (nlocals) {
		if (!( locals = calloc(nlocals, sizeof(locals[0])) )) {
			errobj_set_nomem(interp);
			return NULL;
		}
	} else
		locals = NULL;

	ScopeObject **parents;
	size_t nparents;
	if (parent) {
		nparents = parent->nparents + 1;
		if (!( parents = malloc(nparents * sizeof(parents[0])) )) {
			free(locals);
			errobj_set_nomem(interp);
			return NULL;
		}
	} else {
		parents = NULL;
		nparents = 0;
	}

	ScopeObject *obj = object_new(interp, &scopeobj_type, destroy_scope, sizeof(*obj));
	if (!obj) {
		free(parents);
		free(locals);
		return NULL;
	}

	if (parent) {
		memcpy(parents, parent->parents, parent->nparents * sizeof *parent->parents);
		parents[parent->nparents] = parent;
		OBJECT_INCREF(parent);
	}

	obj->locals = locals;
	obj->nlocals = nlocals;
	obj->parents = parents;
	obj->nparents = nparents;
	return obj;
}

ScopeObject *scopeobj_newglobal(Interp *interp)
{
	ScopeObject *res = scopeobj_newsub(interp, NULL, builtin_nobjects);
	if(!res)
		return NULL;

	memcpy(res->locals, builtin_objects, builtin_nobjects*sizeof(builtin_objects[0]));
	for (size_t i = 0; i < builtin_nobjects; i++)
		OBJECT_INCREF(res->locals[i]);
	return res;
}

ScopeObject *scopeobj_getforlevel(ScopeObject *scope, size_t level)
{
	assert(level <= scope->nparents);
	return (level == scope->nparents) ? scope : scope->parents[level];
}

const struct Type scopeobj_type = { .methods = NULL, .nmethods = 0 };
